set_transformers/tensorflow/layers.py

```diff
@@ -0,0 +1,456 @@
+            MAB kwargs
+        """
+        super().__init__(**kwargs)
+
+        self.k = k
+        self.s = self.add_weight(
+            name="seed_vectors",
+            shape=(k, d),
+            initializer="random_normal",
+            trainable=True
+        )
+
+        self.rff = RFF(
+            **rff_kwargs
+        )
+
+        self.mab = MAB(
+            **mab_kwargs,
+            attention_axes=attention_axes
+        )
+
+    def call(
+        self,
+        z: tf.Tensor
+    ) -> tf.Tensor:
+        """PMA(Z) = MAB(S, rFF(Z)) ∈ R k×d
+        where S ∈ R k×d
+
+        Parameters
+        ----------
+        z : tf.Tensor
+            shape (batch_size, n, d)
+
+        Returns
+        -------
+        tf.Tensor
+            shape (batch_size, k, d)
+        """
+        batch_shape = z.shape[:-2]
+
+        repeated_s = self.s
+        for size in reversed(batch_shape):
+            repeated_s = tf.repeat(
+                tf.expand_dims(
+                    repeated_s,
+                    axis=0
+                ),
+                (size if size is not None else 1,),
+                axis=0
+            )
+        return self.mab(repeated_s, self.rff(z))
+
+
+class Encoder(tf.keras.layers.Layer):
+
+    def __init__(
+        self,
+        type: str,
+        attention_axes: Tuple[int, ...],
+        kwargs_per_layer: List[Dict],
+        **kwargs
+    ):
+        """Set Transformer encoder
+        Stack of SAB or ISAB blocks
+           Encoder(X) = SAB(SAB(... X))
+        or Encoder(X) = ISAB(ISAB(... X))
+
+        Parameters
+        ----------
+        type : str
+            one of ["SAB", "ISAB"]
+        attention_axes : Tuple[int, ...]
+            determines the axes on which attention is performed
+        kwargs_per_layer : List[Dict]
+            kwargs for SAB or ISAB class
+
+        Raises
+        ------
+        AssertionError
+            if type is not in ["SAB", "ISAB"]
+        """
+        super().__init__(**kwargs)
+
+        if type not in ["SAB", "ISAB"]:
+            raise AssertionError(
+                "type should be one of [`SAB`, `ISAB`]"
+            )
+        self.seq = tf.keras.Sequential(
+            layers=[
+                (
+                    SAB
+                    if type == "SAB"
+                    else ISAB
+                )
+                (
+                    **layer_kwargs,
+                    attention_axes=attention_axes
+                )
+                for layer_kwargs in kwargs_per_layer
+            ]
+        )
+
+    def call(
+        self,
+        x: tf.Tensor
+    ) -> tf.Tensor:
+        """Encoder(X) = SAB(SAB(... X))
+        or Encoder(X) = ISAB(ISAB(... X))
+
+        Parameters
+        ----------
+        x : tf.Tensor
+            shape (batch_size, n, d)
+
+        Returns
+        -------
+        tf.Tensor
+            shape (batch_size, n or m, d)
+        """
+        return self.seq(x)
+
+
+class Decoder(tf.keras.layers.Layer):
```

Comment:
IMO, due to its architecture, it should be defined as a function (for example define a function get_set_transformer_decoder) that returns a Sequential model